internal/proxyservice/sshagent/proxy_service.go

```diff
@@ -0,0 +1,162 @@
+		return nil, err
+	}
+
+	return &proxyService{
+		retrieveCredentials: retrieveCredentials,
+		listener:            listener,
+		logger:              logger,
+		done:                false,
+		keyring:             agent.NewKeyring(),
+	}, nil
+}
+
+func parseKey(pemStr []byte) (interface{}, error) {
+	block, _ := pem.Decode(pemStr)
+	if block == nil {
+		return nil, fmt.Errorf("failed to decode PEM block")
+	}
+
+	switch block.Type {
+	case "RSA PRIVATE KEY":
+		return x509.ParsePKCS1PrivateKey(block.Bytes)
+	case "EC PRIVATE KEY":
+		return x509.ParseECPrivateKey(block.Bytes)
+	default:
+		return nil, fmt.Errorf("unsupported key type %q", block.Type)
+	}
+}
+
+// Start initiates the net.Listener to listen for incoming connections
+func (proxy *proxyService) Start() error {
+	logger := proxy.logger
+
+	logger.Infof("Starting service")
+
+	if proxy.done {
+		return fmt.Errorf("cannot call Start on stopped ProxyService")
+	}
+
+	// TODO: this proxy service only fetches credentials at the outset
+	//   so can not cope with rotation
+	backendCredentials, err := proxy.retrieveCredentials()
+	defer internal.ZeroizeCredentials(backendCredentials)
+	if err != nil {
+		return fmt.Errorf("failed on retrieve credentials: %s", err)
+	}
+
+	key := agent.AddedKey{}
+
+	if rsa, ok := backendCredentials["rsa"]; ok {
+		key.PrivateKey, err = parseKey(rsa)
+		if err != nil {
+			return err
+		}
+	}
+	if ecdsa, ok := backendCredentials["ecdsa"]; ok {
+		key.PrivateKey, err = parseKey(ecdsa)
+		if err != nil {
+			return err
+		}
+	}
+
+	// TODO: comment, lifetime and confirm aren't credentials.
+	// Maybe Handler needs a mechanism for these types of non-secret configuration options.
+	if comment, ok := backendCredentials["comment"]; ok {
+		key.Comment = string(comment)
+	}
+	if lifetime, ok := backendCredentials["lifetime"]; ok {
+		var lt uint64
+		lt, err = strconv.ParseUint(string(lifetime), 10, 32)
+		if err != nil {
+			return err
+		}
+		key.LifetimeSecs = uint32(lt)
+	}
+	if confirm, ok := backendCredentials["confirm"]; ok {
+		key.ConfirmBeforeUse, err = strconv.ParseBool(string(confirm))
+		if err != nil {
+			return err
+		}
+	}
+
+	err = proxy.keyring.Add(key)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		for !proxy.done {
+			conn, err := proxy.listener.Accept()
+			if err != nil {
+				logger.Errorf("Failed on accept connection: %s", err)
+				return
+			}
+
+			go func() {
+				proxy.logger.Infof("Serving connection on %v", conn.LocalAddr())
+				err := agent.ServeAgent(proxy.keyring, conn)
+				if err != nil && err != io.EOF {
+					logger.Errorf("Failed on handle connection: %s", err)
+					return
+				}
+
+				logger.Infof("Connection closed on %v", conn.LocalAddr())
+			}()
+		}
+	}()
+
+	return nil
+}
+
+// Stop terminates proxyService by closing the listening net.Listener
+func (proxy *proxyService) Stop() error {
+	proxy.logger.Infof("Stopping service")
+	proxy.done = true
+	return proxy.listener.Close()
+}
```

Comment:
Can add a TODO if you want, but we should verify with a channel that it actually outer goroutine actually stops.

Reply:
TODO